mantidproject/mantid
# Shade Groups When All Sub-Rows Complete In Refl GUI

**Description of work.**
This PR adds a pointer from each row to the parent group to allow the group to update its shade when all its children are complete. That's it. 

So why is it so big?

  Turns out the groups get moved and copied around a lot. This meant that the pointer would often point to the wrong memory address if the group was moved or copied. This had a tendancy to cause weird bugs/segfaults, so new copy and move constructors were created to ensure that groups updated all their children whenever they were moved. 
  The process also included a lot of trial and error. In attempting to remove the copy constructor, a lot of maintenance was done removing unnecessary copies where they occured. We had to put the copy constructor back in the end. But the work done removing the copies was still worth keeping.  

**To test:**
1. Ensure you have the ISIS sample data downloaded
1. Open the ISIS Reflectometry GUI
2. Load a Batch (top left) from the sample data, and click the `+` button to expand all groups
3. Click each run within the first group and click process *one at a time*
4. You should see the highlight colour of the group change when both individual runs are complete
5. On the next group, select one run and process it. 
6. Select the other run and delete that row
7. You should see the group change to a green highlight as all runs within it are now complete.

Fixes #26531 



---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
